### DISCUSSION TITLE
Transaction Advice

### DISCUSSION THREAD

**Discussion by bennythemink** [pocketbase/pocketbase#5265] [Q&A] [answered]
Hello,

I'm having a little difficulty with attempting to get transactions working. I have a route registered and its handler function below. From logging I can see that the problem lies with the 'saveRecord' function call. Everything logs to the console up to the save, but not after. 

The collection API rules are all open, so I do not believe the problem lies there. I've ensured that the collection name is correct. Any pointers would be appreciated.

Many thanks

```js
routerAdd('GET', '/validations/create/:assigner/:assignee/:program_unit', (c) => {
  console.log('IN validations/create')

  let assigner = c.pathParam('assigner')
  let assignee = c.pathParam('assignee')
  let puId = c.pathParam('program_unit')

  console.log('params: ', assigner, assignee, puId)

  try {
    $app.dao().runInTransaction((txDao) => {
      console.log('IN TRANSACTION')

      const stepCollection = $app.dao().findCollectionByNameOrId('validation_steps')
      console.log('stepCollection:', stepCollection)

      const record = new Record(stepCollection)
      record.set('assigner', assigner)
      record.set('assignee', assignee)

      console.log('record:', record)
      $app.dao().saveRecord(record)
      console.log('record saved')

      return c.json(200, { message: 'saved' })
    })
  } catch (err) {
    console.log('ERROR:', err)
  }

  return c.json(200, { message: 'Hello ' + assigner + ' and ' + assignee })
})
```

**Comment by ganigeorgiev** [comment, marked as answer]
When starting a transaction you have to use the `Dao` from the callback argument otherwise it could cause deadlock because you can have only a single writer/transaction at a time, aka. replace `$app.dao()` with `txDao` inside of the transaction.

As a general advice - avoid transactions (especially mixed read&write) if you don't need them as they  could cause a performance bottleneck.

I don't understand the goal of the above code sample to provide a proper recommendation (why do you call `c.json` from inside the transaction? or why there is even need for a transaction in the first place?) but note that calling `saveRecord` doesn't perform validations (aka. it roughly translates to a direct `INSERT`/`UPDATE` staments) and if you are operating with untrusted user input you have to use the `RecordUpsertForm` (see [Creating new record WITH data validations](https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations); when using the form inside of a transaction you'll have to also set its dao with `form.setDao(txDao)`).
This is simplified with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and the form will be no longer needed.

**Reply by bennythemink** [reply]
Thanks @ganigeorgiev,

That was exactly it, my use of the `$app.dao()` was causing a deadlock. The code I provided was just my sample code to get a transaction working. My actual use case has a number of writes (and a read) that needs to be transaction based, thankfully all data going into the database in this instance is non user inputted but will keep in mind the `RecordUpsertForm`.

Thanks for the advice, I will avoid transactions where possible. Looking forward to the refactoring, it sounds good!
